### PR TITLE
Minimal changes for CPU CropMirrorNormalize

### PR DESCRIPTION
**Pull request by pribalta**
Minimal changes set of changes from https://github.com/NVIDIA/DALI/pull/187 that enable CropMirrorNormalize on CPU.

### CODE REVIEW

File: dali/pipeline/operators/fused/crop_mirror_normalize.cc [hunk 1]
```diff
@@ -14,52 +14,188 @@
...
+      }
+    }
+  } else {
+    for (int tid = 0; tid < nStride; ++tid) {
+      const int c = tid % pad_C;
+      const int w = (tid / pad_C) % W;
+      const int h = tid / (pad_C * W);
+
+      float input;
+      if (pad && c == 3) {
```
Comment (JanuszL) [inline dali/pipeline/operators/fused/crop_mirror_normalize.cc]: pad is already false here.

File: dali/pipeline/operators/fused/crop_mirror_normalize.cc [hunk 2]
```diff
@@ -14,52 +14,188 @@
...
+    const int W,
+    const bool pad,
+    const int mirror_image,
+    const float* mean,
+    const float* inv_std,
+    const uint8* input_ptr,
+    const int in_step,
+    DALITensorLayout layout,
+    Out* output_ptr) {
+  const int pad_C = pad ? 4 : C;
```
[the remark this answers is not in the document]
Comment (pribalta) [reply dali/pipeline/operators/fused/crop_mirror_normalize.cc]: accepted

File: dali/pipeline/operators/fused/crop_mirror_normalize.cc (hunk 1 above)
Comment (pribalta) [reply dali/pipeline/operators/fused/crop_mirror_normalize.cc]: Here we are on the case where layout is HWC, so pad can actually be true or false in this statement

Comment (JanuszL) [reply dali/pipeline/operators/fused/crop_mirror_normalize.cc]: You are right, I missed that this else refers to the different if than I thought.

**Comment by JanuszL** [comment]
Please add also coauthorship of @drivanov to the commit message.

File: dali/pipeline/operators/fused/crop_mirror_normalize.cc [hunk 3]
```diff
@@ -14,52 +14,188 @@
...
+    const float* inv_std,
+    const uint8* input_ptr,
+    const int in_step,
+    DALITensorLayout layout,
+    Out* output_ptr) {
+  const int pad_C = C;
+  const int nStride = pad_C*H*W;
+
+  const int a = mirror_image? (W - 1) * C : 0;
+  const int b = mirror_image? -C : C;
```
Comment (klecki) [inline dali/pipeline/operators/fused/crop_mirror_normalize.cc]: Could you use better names for those variables? As far as it goes, `h`, `w`, `c` (and `H`, `W`, `C`) are ok, but `a` and `b` don't have any meaning.

File: dali/pipeline/operators/fused/crop_mirror_normalize.cc [hunk 4]
```diff
@@ -14,52 +14,188 @@
...
+  if (layout == DALI_NCHW) {
+    // Coalesced writes
+    for (int c=0; c < C; ++c) {
+      for (int h=0; h < H; ++h) {
+        for (int w=0; w < W; ++w) {
+          const int in_idx = a + c + b * w + in_step*h;   // HWC
+          const int out_idx = (c*H + h)*W + w;            // CHW
+
+          output_ptr[out_idx] = static_cast<Out>(
+              (static_cast<float>(input_ptr[in_idx])-mean[c]) * inv_std[c]);
```
Comment (klecki) [inline dali/pipeline/operators/fused/crop_mirror_normalize.cc]: As `mean` and `inv_std` are already floats, the `static_cast<float>` is unnecessary.

File: dali/pipeline/operators/fused/crop_mirror_normalize.cc [hunk 5]
```diff
@@ -14,52 +14,188 @@
...
+        }
+      }
+    }
+
+    // Pad to 4 channels with 0s
+    if (pad) {
+      const Out out = static_cast<Out>(0);
+      for (int c=C; c < 4; ++c) {
```
Comment (klecki) [inline dali/pipeline/operators/fused/crop_mirror_normalize.cc]: I would guess, that if we want to be efficient, we should merge the padding part with regular data processing loop above.

We don't want to have if(pad) in innermost loop, so I would suggest to do something like:
```
if (NCHW) {
    if (pad) {
        nchw_pad_loop();
    } else {
        nchw_no_pad_loop();
   }
}
```
And so on. But I am not sure if moving the loops out will actually improve readability.

File: dali/pipeline/operators/fused/crop_mirror_normalize.cc [hunk 6]
```diff
@@ -14,52 +14,188 @@
...
+        for (int h=0; h < H; ++h) {
+          for (int w=0; w < W; ++w) {
+            const int out_idx = (c*H + h)*W + w;  // CHW
+            output_ptr[out_idx] = out;
+          }
+        }
+      }
+    }
+  } else {
+    for (int tid = 0; tid < nStride; ++tid) {
```
Comment (klecki) [inline dali/pipeline/operators/fused/crop_mirror_normalize.cc]: Why not just regular three-times nested loop? We go NHWC -> NHWC here, and that would get rid of all those divisions and modulos below.

File: dali/pipeline/operators/fused/crop_mirror_normalize.cc [hunk 7]
```diff
@@ -14,52 +14,188 @@
...
+      const int c = tid % pad_C;
+      const int w = (tid / pad_C) % W;
+      const int h = tid / (pad_C * W);
+
+      float input;
+      if (pad && c == 3) {
```
Comment (klecki) [inline dali/pipeline/operators/fused/crop_mirror_normalize.cc]: This probably won't work for GRAY: we will never get `c == 3`, as `pad_C = 1`.

File: dali/pipeline/operators/fused/crop_mirror_normalize.cc (hunk 3 above)
Comment (pribalta) [reply dali/pipeline/operators/fused/crop_mirror_normalize.cc]: accepted

File: dali/pipeline/operators/fused/crop_mirror_normalize.cc [hunk 8]
```diff
@@ -14,52 +14,188 @@
...
+    const int in_step,
+    DALITensorLayout layout,
+    Out* output_ptr) {
+  const int pad_C = C;
+  const int nStride = pad_C*H*W;
+
+  const int a = mirror_image? (W - 1) * C : 0;
+  const int b = mirror_image? -C : C;
+  if (layout == DALI_NCHW) {
+    // Coalesced writes
```
[the remark this answers is not in the document]
Comment (pribalta) [reply dali/pipeline/operators/fused/crop_mirror_normalize.cc]: accepted

File: dali/pipeline/operators/fused/crop_mirror_normalize.cc (hunk 4 above)
Comment (pribalta) [reply dali/pipeline/operators/fused/crop_mirror_normalize.cc]: accepted

File: dali/pipeline/operators/fused/crop_mirror_normalize.cc (hunk 5 above)
Comment (JanuszL) [reply dali/pipeline/operators/fused/crop_mirror_normalize.cc]: No, fusing with above loop wont help. Channel loop is outermost so having this as
```
for (0 to c) {...}
for (c to 4) {...}
```
is the same as
```
for (0 to 4) {
    if (c < 4) {...}
    else {...}
}
```

Comment (pribalta) [reply dali/pipeline/operators/fused/crop_mirror_normalize.cc]: I think that I will not be getting into that at the moment. I'd rather treat the optimization of this kernel as a separate item

File: dali/pipeline/operators/fused/crop_mirror_normalize.cc (hunk 6 above)
Comment (pribalta) [reply dali/pipeline/operators/fused/crop_mirror_normalize.cc]: Probably this change should be undertaken with a wider scope for this kernel, as there could be several things to improve. By now i'll roll over these suggestions to a different item of work

File: dali/pipeline/operators/fused/crop_mirror_normalize.cc (hunk 7 above)
Comment (pribalta) [reply dali/pipeline/operators/fused/crop_mirror_normalize.cc]: I will try to clarify this, but what I understand from a comment in some lines above is that padding seems to be considered only when we have 3 channels

**Comment by pribalta** [comment]
builds 39601

**Pull request merged**